README.md

```diff
@@ -13,3 +15,25 @@
   complexity suggests a different time.
 
 Add your answers to this markdown file.
+
+## Solution
+
+**_List 3 reasons why asymptotic analysis may be misleading with respect to actual performance in practice._**
+
+1. Asymptotic analysis focuses on very large input sizes. This means input sizes which approach infinity, but in practice input sizes can be much smaller in alot of situations. An algorithm with better asymptotic complexity may have larger overhead that makes it perform worse for smaller input sizes. This is clear when comparing the graph of diferent time complexities. A polynomial or linear algorithm can perform better with a smaller input size then its logarithmic counterpart. Then as the input gets larger it will change and now the logarithmic complexity will be best.
+
+2. Asymptotic analysis ignores constants. In reality, constants matter and can significantly impact actual running times. Say we have a algorithm which has 100 constant operations it must perform. Some of these constant factors could also be doing alot of work where they themselfs do alot of constant operations. When we ignore the lower order terms it may be misleading for certain algorithms.
+
+3. Sometimes the underlying hardware could be an issue when comparing the real world performance to theoretical ones. For example diferent cache and branch-predicitions may mean a asympototicaly worse algorithm can perform better than one which is supposed to be superior. This can depend on the real-world hardware and can change between machines with diferent architechtures.
+
+**_Suppose finding a particular element in a binary search tree with 1,000 elements takes 5 seconds. Given what you know about the asymptotic complexity of search in a binary search tree, how long would you guess finding the same element in a search tree with 10,000 elements takes? Explain your reasoning_**
+
+Searching a binary-tree has a average complexity of $O(\log n)$. Furthermore going from a input size of 1,000 to 10,000 elements increases the input size by a factor of 10. When we compare $log(1,000)$ to $log(10,000)$ we see that $log(1,000) = 10$ and $log(10,000) = 13.33$. So I would expect searching 10,000 elements to take roughly $13.3 / 10 = 1.33$ times as long as searching 1,000 elements. If searching 1,000 elements takes 5 seconds, I would then predict searching 10,000 elements would take about 6.65 seconds.
+
+**_You measure the time with 10,000 elements and it takes 100 seconds! List 3 reasons why this could be the case, given that reasoning with the asymptotic complexity suggests a different time_**
+
+1. If the tree is unballanced aka like a linear **linked-list** then it would be $O(n)$ search time instead of $O(\log n)$.
```

Comment:
...and could that account for a difference this large?

Reply:
It would add a 10x time. However this along with other things listed below could easily take it over the expected time to 100s.

These things could ALL be factors not just one. If the binary tree is unballanced then it would mean a 10x slowdown from that alone. Then the other factors could easily explain the 100s time.

Reply:
The reasons have to be valid on their own.